agent/consul/state/stream_publisher.go

```diff
@@ -0,0 +1,439 @@
+				// and effectively re-start all subscriptions to be on the safe side
+				// without just crashing?
+				// TODO(banks): reset all instead of panic?
+				panic(err)
+			}
+
+			continue
+		}
+
+		// Split events by topic to deliver.
+		eventsByTopic[event.Topic] = append(eventsByTopic[event.Topic], event)
+	}
+
+	// Deliver events
+	for topic, events := range eventsByTopic {
+		buf, ok := e.topicBuffers[topic]
+		if !ok {
+			buf = stream.NewEventBuffer()
+			e.topicBuffers[topic] = buf
+		}
+		buf.Append(events)
+	}
+}
+
+// handleACLUpdate handles an ACL token/policy/role update. This method assumes
+// the lock is held.
+func (e *EventPublisher) handleACLUpdate(tx *txnWrapper, event agentpb.Event) error {
+	switch event.Topic {
+	case agentpb.Topic_ACLTokens:
+		token := event.GetACLToken()
+		subs := e.subsByToken[secretHash(token.Token.SecretID)]
+		for _, sub := range subs {
+			sub.CloseReload()
+		}
+	case agentpb.Topic_ACLPolicies:
+		policy := event.GetACLPolicy()
+		// TODO(streaming) figure out how to thread method/ent meta here for
+		// namespace support in Ent. Probably need wildcard here?
+		tokens, err := e.store.aclTokenListByPolicy(tx, policy.PolicyID, nil)
+		if err != nil {
+			return err
+		}
+
+		// Loop through the tokens used by the policy.
+		for token := tokens.Next(); token != nil; token = tokens.Next() {
+			token := token.(*structs.ACLToken)
+			if subs, ok := e.subsByToken[secretHash(token.SecretID)]; ok {
+				for _, sub := range subs {
+					sub.CloseReload()
+				}
+			}
+		}
+
+		// Find any roles using this policy so tokens with those roles can be reloaded.
+		roles, err := e.store.aclRoleListByPolicy(tx, policy.PolicyID, nil)
+		if err != nil {
+			return err
+		}
+		for role := roles.Next(); role != nil; role = roles.Next() {
+			role := role.(*structs.ACLRole)
+
+			// TODO(streaming) figure out how to thread method/ent meta here for
+			// namespace support in Ent.
+			tokens, err := e.store.aclTokenListByRole(tx, role.ID, nil)
+			if err != nil {
+				return err
+			}
+			for token := tokens.Next(); token != nil; token = tokens.Next() {
+				token := token.(*structs.ACLToken)
+				if subs, ok := e.subsByToken[secretHash(token.SecretID)]; ok {
+					for _, sub := range subs {
+						sub.CloseReload()
+					}
+				}
+			}
+		}
+
+	case agentpb.Topic_ACLRoles:
+		role := event.GetACLRole()
+		// TODO(streaming) figure out how to thread method/ent meta here for
+		// namespace support in Ent.
+		tokens, err := e.store.aclTokenListByRole(tx, role.RoleID, nil)
+		if err != nil {
+			return err
+		}
+		for token := tokens.Next(); token != nil; token = tokens.Next() {
+			token := token.(*structs.ACLToken)
+			if subs, ok := e.subsByToken[secretHash(token.SecretID)]; ok {
+				for _, sub := range subs {
+					sub.CloseReload()
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// secretHash returns a 256-bit Blake2 hash of the given string.
+func secretHash(token string) string {
+	hash, err := blake2b.New256(nil)
```

Comment:
I'd like to add more to this docstring to explain why the token is being hashed. 

It is my understanding that Go is already going to hash the key used in a map. I don't think i understand why this is used to create map keys.

Reply:
Yeah a comment would help here. The reason was that originally we were creating a map keyed by token and there was a concern that we were copying secrets around in memory unnecessarily which made them easier to accidentally leak in a debug message or similar.

I'm not super sure if that is relevant now since I _think_ a `string` key will end up being a pointer to the actual string literal that is already in memory in the state store anyway. It's also true that we store secrets in other structures currently - both in ACL caches and in the agent cache keys so on reflection I'm fine with removing it as I'm not sure it it a real security improvement within any realistic threat model. Arbitrary memory access on a server is already outside our threat model and the chances of us accidentally dumping this structure seem extremely low.